activerecord/lib/active_record/delay_touching/state.rb

```diff
@@ -0,0 +1,71 @@
+module ActiveRecord
+  # = Active Record Delay Touching
+  module DelayTouching
+
+    # Tracking of the touch state. This class has no class-level data, so you can
+    # store per-thread instances in thread-local variables.
+    class State # :nodoc:
```

Comment:
@kaspth I don't think a linked list is necessarily better in this case. There are a couple of reasons:

1. The code might exit and re-enter nested transactions. If I understand your suggestion correctly, a linked list would end up having to have branches:

    ```ruby
    transaction do
      transaction do
      end
      transaction do
      end
    end
    ```
2. It's true that `@already_updated_records` is only necessary when there's just one `State` left on the stack, because that's when we start touching records and keeping track of those touches. However, if I move it out to another instance of `State` (which I think is what you're suggesting), it's harder to keep track of. The `touch` override has to check it, which requires storing it as another thread-local variable. I think it will get messy.

As I wrote this answer, though, I did begin to realize that at the very least, I should rename `@already_updated_records` to `@already_touched_records`. But I realize that doesn't address your concerns about keeping track of two states at once.